mlfinlab/data_structures/imbalance_data_structures.py

```diff
@@ -0,0 +1,287 @@
+
+        imb_flag = False
+        if len(imbalance_array) < exp_num_ticks:
+            exp_tick_imb = np.nan  # waiting for array to fill for ewma
+        else:
+            # expected imbalance per tick
+            ewma_window = int(exp_num_ticks * num_prev_bars)
+            exp_tick_imb = ewma(
+                np.array(imbalance_array[-ewma_window:], dtype=float), window=ewma_window)[-1]
+            if np.abs(theta_imb) > exp_num_ticks * np.abs(exp_tick_imb):
+                imb_flag = True
+
+        # Check min max
+        if price > high_price:
+            high_price = price
+        if price <= low_price:
+            low_price = price
+
+        prev_price = price  # update prev_price
+        prev_tick_rule = tick_rule
+
+        #print(date_time, cum_dollar_imb, exp_num_ticks, exp_tick_imb)
+        # If threshold reached then take a sample
+        if imb_flag is True:   # pylint: disable=eval-used
+            # Create bars
+            open_price = cache[0][1]
+            low_price = min(low_price, open_price)
+            close_price = price
+            num_ticks_bar.append(cum_ticks)
+            expected_num_ticks_bar = ewma(
+                np.array(num_ticks_bar[-num_ticks_ewma_window:], dtype=float), num_ticks_ewma_window)[-1]  # expected number of ticks based on formed bars
+            # Update bars & Reset counters
+            list_bars.append([date_time, open_price, high_price, low_price, close_price,
+                              cum_volume, cum_dollar_value, cum_ticks])
+            cum_ticks, cum_dollar_value, cum_volume, cum_dollar_imb, cum_tick_imb, cum_volume_imb = 0, 0, 0, 0, 0, 0
+            high_price, low_price = -np.inf, np.inf
+            exp_num_ticks = expected_num_ticks_bar
+
+        # Update cache
+        cache.append([date_time, price, low_price, high_price,
+                      cum_volume, cum_dollar_value, cum_ticks, cum_dollar_imb, cum_tick_imb, cum_volume_imb, exp_num_ticks, imbalance_array])
+    return list_bars, cache, num_ticks_bar, prev_price, prev_tick_rule
+
+
+def _assert_dataframe(test_batch):
+    """
+    Tests that the csv file read has the format: date_time, price, & volume.
+    If not then the user needs to create such a file. This format is in place to remove any unwanted overhead.
+
+    :param test_batch: DataFrame which will be tested.
+    """
+    assert test_batch.shape[1] == 3, 'Must have only 3 columns in csv: date_time, price, & volume.'
+    assert isinstance(test_batch.iloc[0, 1],
+                      float), 'price column in csv not float.'
+    assert isinstance(test_batch.iloc[0, 2],
+                      np.int64), 'volume column in csv not int.'
+
+    try:
+        pd.to_datetime(test_batch.iloc[0, 0])
+    except ValueError:
+        print('csv file, column 0, not a date time format:',
+              test_batch.iloc[0, 0])
+
+
+def _batch_run(file_path, metric, exp_num_ticks_init, num_prev_bars, num_ticks_ewma_window, batch_size=20000000):
```

Comment:
It looks better, IMO :) But this is just a suggestion
```suggestion
def _batch_run(file_path, metric, exp_num_ticks_init, num_prev_bars, num_ticks_ewma_window, batch_size=2e7):
```